src/merkle.rs

```diff
@@ -58,15 +58,66 @@ impl<'p, D: KeyValueDB, P: PoolParams> MerkleTree<'p, D, P> {
         I::IntoIter: 'a,
         P::Fr: 'a,
     {
-        let mut batch = self.db.transaction();
-
         for (index, hash, temporary) in hashes.into_iter().cloned() {
+            let mut batch = self.db.transaction();
             self.add_hash_batched(&mut batch, index, hash, temporary);
+            self.db.write(batch).unwrap();
         }
 
+        // self.cleanup();
+    }
+
+    pub fn add_subtree(&mut self, hashes: &Vec<Hash<P::Fr>>, start_index: u32, temporary: bool)
+    {
+        let size = hashes.len();
+
+        assert!((size & (size - 1)) == 0, "subtree size should be a power of 2");
+
+        let mut batch = self.db.transaction();
+
+        // set leaves
+        for index_shift in 0..size {
+            let index = start_index + index_shift as u32;
+
+            // println!("copy {} {} {}", 0, index, hashes[index_shift]);
+            self.set_batched(&mut batch, 0, index, hashes[index_shift]);
+
+            if !temporary {
+                // mark this hash as non-removable for later
+                self.set_retained_batched(&mut batch, index, hashes[index_shift]);
+            }
+        }
+
+        // build subtree
+        let mut child_hashes = hashes.clone();
```

Comment:
This code can be rewritten without extra Vec allocations, but I'm not sure what is faster: allocating and indexing through vecs or not allocating and using Database::get. We should profile it later.

For now, pre-allocate `parent_hashes` with capacity.